### PR TITLE
build DiscordClient on top of IOC

### LINKED ISSUE #1683: Centralized ServiceProvider
# Summary 
We want to get away from individual ServiceProviders per extension and when we have the ServiceProvider in the DiscordClient we can start to use DI on AsyncEvents

### PR DESCRIPTION
**Pull request by akiraveliara**
current status:
- [x] RestClient capable of construction
- [x] Gateway incapable of construction
  - [x] dispatch constructed and extensible through IOC
  - [x] possibly the raw gw client extensible through IOC?
- [x] expand capability of DiscordClientBuilder to allow configuring logging and services
- [x] process events through DiscordClientBuilder
- [x] obsolete old events (on DiscordClient only, since this won't support sharding yet)
- [x] docs

non-goals of this PR:
- sharded client support (future PR)
- extension support (future PR)
- exposing more extensibility interfaces (future PR, open issues for each)
- ratelimiting constructed and extensible through IOC (future PR)

leave issues for any other features associated with this you may want

closes #1683
makes progress towards #1820

### CODE REVIEW

File: DSharpPlus/Extensions/ServiceCollectionExtensions.InternalSetup.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+public static class ServiceCollectionExtensions
+{
+    internal static IServiceCollection AddDSharpPlusDefaultsSingleShard(this IServiceCollection serviceCollection)
+    {
+        // peripheral setup
+        serviceCollection.AddSingleton<IMessageCacheProvider, MessageCache>();
+
+        // rest setup
+        serviceCollection.AddKeyedSingleton<HttpClient>("DSharpPlus.Rest.HttpClient")
```
Comment (VelvetToroyashi) [inline DSharpPlus/Extensions/ServiceCollectionExtensions.InternalSetup.cs:19]: Consider a constants class

File: DSharpPlus/Net/Rest/RestClient.cs [hunk 2]
```diff
@@ -58,21 +69,10 @@ internal RestClient
...
-        };
-
-        this.httpClient = new HttpClient(httphandler)
-        {
-            BaseAddress = new Uri(Utilities.GetApiBaseUri()),
-            Timeout = timeout
-        };
-
+        this.httpClient.BaseAddress = new Uri(Utilities.GetApiBaseUri());
+        this.httpClient.Timeout = timeout;
```
Comment (VelvetToroyashi) [inline DSharpPlus/Net/Rest/RestClient.cs:75]: Perhaps it'd be worthwhile to go the Remora route and expose a [rest] client builder; there's occasional reason to change what the client does (e.g., adding additional polly handlers for metrics, or using the canary api for one reason or another)

**Review by VelvetToroyashi — COMMENTED** [review]
LGTM aside from comments; gonna open an issue wrt some tomfoolery happening regarding ratelimit handling

File: DSharpPlus/Net/Rest/RestClient.cs (hunk 2 above)
Comment (VelvetToroyashi) [reply DSharpPlus/Net/Rest/RestClient.cs:75]: (Client builder as in `Action<HttpClient>` parameter)

File: DSharpPlus/Extensions/ServiceCollectionExtensions.InternalSetup.cs (hunk 1 above)
Comment (akiraveliara) [reply DSharpPlus/Extensions/ServiceCollectionExtensions.InternalSetup.cs:23]: no

File: DSharpPlus/Net/Rest/RestClient.cs (hunk 2 above)
Comment (akiraveliara) [reply DSharpPlus/Net/Rest/RestClient.cs:75]: i think this is something we should look into in a separate pull request to customize specifically rest further

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 3]
```diff
@@ -671,20 +673,26 @@ internal async Task OnReadyEventAsync(ReadyPayload ready, JArray rawGuilds, JArr
             this.guilds[guild.Id] = guild;
         }
 
-        await this.ready.InvokeAsync(this, new SessionReadyEventArgs());
+        await this.events[typeof(SessionCreatedEventArgs)]
+            .As<SessionCreatedEventArgs>()
+            .InvokeAsync(this, new SessionCreatedEventArgs());
```
Comment (OoLunar) [inline DSharpPlus/Clients/DiscordClient.Dispatch.cs:678]: So... every time a gateway event is received, a dictionary lookup is required with a type cast?

File: DSharpPlus/Clients/DiscordClient.Events.cs [hunk 4]
```diff
@@ -1,49 +1,48 @@
...
 
 namespace DSharpPlus;
 
 public sealed partial class DiscordClient
 {
-    internal static TimeSpan EventExecutionLimit { get; } = TimeSpan.FromSeconds(1);
-
     // oh lord why did you have to pack into regions
     // this makes simple copy-paste ineffective
     // :notlikethis:
```
Comment (OoLunar) [inline DSharpPlus/Clients/DiscordClient.Events.cs:12]: What regions

File: DSharpPlus/Clients/DiscordClient.cs [hunk 5]
```diff
@@ -25,19 +31,36 @@ namespace DSharpPlus;
...
+    private readonly IClientErrorHandler errorHandler;
 
     private List<BaseExtension> extensions = [];
-    private StatusUpdate status = null;
+    private StatusUpdate? status = null;
 
-    private ManualResetEventSlim ConnectionLock { get; } = new ManualResetEventSlim(true);
+    private int heartbeatInterval;
+    private DateTimeOffset lastHeartbeat;
+    private int skippedHeartbeats = 0;
```
Comment (OoLunar) [inline DSharpPlus/Clients/DiscordClient.cs:48]: Why is this explicitly initiated to it's default value?

**Review by OoLunar — APPROVED** [review]
I think this is okay? Regardless please write a migration guide for this PR specifically. Before we release v5, we'll merge the migration guide for this PR and all other breaking changes into one. I just want a guide to reference right now since I myself don't particularly understand what the intended use is.

File: DSharpPlus/Clients/DiscordClient.Events.cs (hunk 4 above)
Comment (akiraveliara) [reply DSharpPlus/Clients/DiscordClient.Events.cs:12]: the ones i removed.

File: DSharpPlus/Clients/DiscordClient.cs (hunk 5 above)
Comment (akiraveliara) [reply DSharpPlus/Clients/DiscordClient.cs:48]: ask emzi, i can remove it

Comment (VelvetToroyashi) [reply DSharpPlus/Clients/DiscordClient.cs:48]: Something you'll have to ask Emzi...

Comment (OoLunar) [reply DSharpPlus/Clients/DiscordClient.cs:48]: Just remove it, it's fine

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs (hunk 3 above)
Comment (OoLunar) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:678]: See https://discord.com/channels/379378609942560770/379386901725052928/1245887242824847422

**Review by VelvetToroyashi — APPROVED** [review]
Cursory glance (namely because of a jumble of a diff), but LGTM.

**Comment by akiraveliara** [comment]
@OoLunar added the migration guide as requested

File: docs/articles/migration/4x_to_5x.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+> The API surface of DSharpPlus 5.x is not stable yet. This migration guide may be incomplete or outdated. It is recommended to cross-reference with the [tracking issue](https://github.com/DSharpPlus/DSharpPlus/issues/1585) when migrating.
+
+The first change you will likely encounter is a rewrite to how bots are set up. We now support two approaches to the problem instead of the old approach:
+
+# [DiscordClientBuilder](#tab/discordclientbuilder)
+
+The simplest way to get a bot running is to use `DSharpPlus.DiscordClientBuilder`. To get started, create a new builder as follows: 
+
+```cs
+DiscordClientBuilder builder = DiscordClientBuilder.Default(string token, DiscordIntents intents);
```
Comment (OoLunar) [inline docs/articles/migration/4x_to_5x.md:18]: Perhaps the method should be called `CreateDefault`?

**Review by OoLunar — APPROVED** [review]

File: docs/articles/migration/4x_to_5x.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+uid: articles.migration.4x_to_5x
+title: Migration 4.x - 5.x
+---
+
+## Migration from 4.x to 5.x
+
+> ![NOTE]
+> The API surface of DSharpPlus 5.x is not stable yet. This migration guide may be incomplete or outdated. It is recommended to cross-reference with the [tracking issue](https://github.com/DSharpPlus/DSharpPlus/issues/1585) when migrating.
+
+The first change you will likely encounter is a rewrite to how bots are set up. We now support two approaches to the problem instead of the old approach:
```
[the remark this answers is not in the document]
Comment (akiraveliara) [reply docs/articles/migration/4x_to_5x.md:11]: the problem of creating a DiscordClient

File: docs/articles/migration/4x_to_5x.md (hunk 6 above)
Comment (akiraveliara) [reply docs/articles/migration/4x_to_5x.md:18]: we could, i don't have any strong opinions on it

**Pull request merged**